compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/ClassReference.kt

```diff
@@ -0,0 +1,162 @@
+    public val clazz: KtClassOrObject,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  public class Descriptor(
+    public val clazz: ClassDescriptor,
+    override val fqName: FqName
+  ) : ClassReference()
+
+  override fun toString(): String {
+    return "${this::class.qualifiedName}($fqName)"
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (other !is ClassReference) return false
+
+    if (fqName != other.fqName) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    return fqName.hashCode()
+  }
+}
+
+/**
+ * Attempts to resolve the [fqName] to a [ClassDescriptor] first, then falls back to a
+ * [KtClassOrObject] if the descriptor resolution fails. This will happen if the code being parsed
+ * was generated as part of the compilation round for this module.
+ */
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.classReferenceOrNull(fqName: FqName): ClassReference? {
+  return fqName.classDescriptorOrNull(this)
+    ?.let { Descriptor(it, fqName) }
+    ?: getKtClassOrObjectOrNull(fqName)
+      ?.let { Psi(it, fqName) }
+}
+
+@ExperimentalAnvilApi
+public fun ClassDescriptor.toClassReference(): Descriptor {
+  return ClassReference.Descriptor(this, fqNameSafe)
+}
+
+@ExperimentalAnvilApi
+public fun KtClassOrObject.toClassReference(): Psi {
+  return ClassReference.Psi(this, requireFqName())
+}
+
+@ExperimentalAnvilApi
+public fun ModuleDescriptor.requireClassReference(fqName: FqName): ClassReference {
+  return classReferenceOrNull(fqName)
+    ?: throw AnvilCompilationException("Couldn't resolve ClassReference for $fqName")
+}
+
+/**
+ * This will return all super types as [ClassReference], whether they're parsed as [KtClassOrObject]
+ * or [ClassDescriptor]. This will include generated code, assuming it has already been generated.
+ * The returned sequence will be distinct by FqName, and Psi types are preferred over Descriptors.
+ *
+ * The first elements in the returned sequence represent the direct superclass to the receiver. The
+ * last elements represent the types which are furthest up-stream.
+ *
+ * @param includeSelf If true, the receiver class is the first element of the sequence
+ */
+@ExperimentalAnvilApi
+public fun ClassReference.allSuperTypeClassReferences(
+  module: ModuleDescriptor,
+  includeSelf: Boolean = false
+): Sequence<ClassReference> {
+
+  fun KtClassOrObject.superTypeEntryNames() = superTypeListEntries
+    .map { it.requireFqName(module) }
+
+  fun List<FqName>.typeRefs(): Sequence<ClassReference> {
+    return asSequence().map { module.requireClassReference(it) }
+  }
+
+  val initial = if (includeSelf) {
+    sequenceOf(this)
+  } else {
+    when (this) {
+      is Descriptor -> {
+        clazz.directSuperClassAndInterfaces()
+          .map { Descriptor(it, it.fqNameSafe) }
+          .asSequence()
+      }
+      is Psi -> {
+        clazz.superTypeEntryNames()
+          .typeRefs()
+      }
+    }
+  }
```

Comment:
This duplication can be likely avoided if you use `sequenceOf(this).drop(if (includeSelf) 0 else 1)`. Or am I missing something?